pkg/compose/cp.go

```diff
@@ -96,6 +87,32 @@ func (s *composeService) Copy(ctx context.Context, projectName string, options a
 	return g.Wait()
 }
 
+func (s *composeService) listContainersTargetedForCopy(ctx context.Context, projectName string, index int, direction copyDirection, serviceName string) (Containers, error) {
+	var containers Containers
+	var err error
+	switch {
+	case direction == fromService && index == 0:
+		containers, err = s.getContainers(ctx, projectName, oneOffExclude, true, serviceName)
+		return containers[:0], err
+	case direction == fromService || (direction == toService && index > 0):
+		container, err := s.getSpecifiedContainer(ctx, projectName, oneOffExclude, true, serviceName, index)
+		if err != nil {
+			return nil, err
+		}
+		return append(containers, container), nil
+	default:
+		containers, err = s.getContainers(ctx, projectName, oneOffExclude, true, serviceName)
```

Comment:
Looks like the first and last branch are basically "the same", so this could be;


```go
	switch {
	case index > 0:
		container, err := s.getSpecifiedContainer(ctx, projectName, oneOffExclude, true, serviceName, index)
		if err != nil {
			return nil, err
		}
		return Containers{container}, nil
	default:
		containers, err := s.getContainers(ctx, projectName, oneOffExclude, true, serviceName)
		if err != nil {
			return nil, err
		}
		if len(containers) < 1 {
			return nil, fmt.Errorf("no container found for service %q", serviceName)
		}
		if direction == fromService {
			// copy from container defaults to the first container
			return containers[:1], nil
		}
		return containers, nil
	}
```

Reply:
I prefer to keep the `s.getContainers()` as it because it's wildly used in the rest of the code. Having a bit more complexity in the `copyFrom` and `copyTo` seems to be a good tradeoff IHMO.

I updated the switch as you proposed 👍